File: src/main/java/ltdjms/discord/aiagent/services/AgentConfigCacheInvalidationListener.java
package ltdjms.discord.aiagent.services;

import java.util.function.Consumer;

import ltdjms.discord.shared.cache.CacheService;
import ltdjms.discord.shared.events.AIAgentChannelConfigChangedEvent;
import ltdjms.discord.shared.events.DomainEvent;

/**
 * AI Agent 頻道配置變更事件的監聽器。
 *
 * <p>當配置變更時，自動使相關快取失效。
 */
public class AgentConfigCacheInvalidationListener implements Consumer<DomainEvent> {

  private final CacheService cacheService;

  /** 快取鍵格式 */
  private static final String CACHE_KEY_FORMAT = "ai:agent:config:%d:%d";

  /**
   * 建立監聽器。
   *
   * @param cacheService 快取服務
   */
  public AgentConfigCacheInvalidationListener(CacheService cacheService) {
    this.cacheService = cacheService;
  }

  @Override
  public void accept(DomainEvent event) {
    if (event instanceof AIAgentChannelConfigChangedEvent e) {
      handleConfigChanged(e);
    }
  }

  /**
   * 處理配置變更事件。
   *
   * @param event 配置變更事件
   */
  private void handleConfigChanged(AIAgentChannelConfigChangedEvent event) {
    String cacheKey = String.format(CACHE_KEY_FORMAT, event.guildId(), event.channelId());
    cacheService.invalidate(cacheKey);
  }
}

File: packages/ai/src/services/routing/agent-config-cache-invalidation-listener.ts
import { type DomainEvent, type DomainEventPublisher, type CacheService } from '@ltdjms/shared';
import pino from 'pino';

const CACHE_KEY_PREFIX = 'ai:agent:config:';

/**
 * Listens for AIAgentChannelConfigChanged events and invalidates the
 * corresponding cache entry so subsequent reads are served from the DB.
 */
export class AgentConfigCacheInvalidationListener {
  private readonly logger: pino.Logger;
  private eventHandler: ((event: DomainEvent) => void) | null = null;

  constructor(
    private readonly cacheService: CacheService,
    private readonly eventPublisher: DomainEventPublisher,
    logger?: pino.Logger,
  ) {
    this.logger = logger ?? pino({ name: 'agent-config-cache-invalidation-listener' });
  }

  register(): void {
    if (this.eventHandler) {
      return;
    }

    this.eventHandler = (event) => {
      const candidate = event as unknown as Record<string, unknown>;
      if (
        candidate.eventType !== 'ai_agent_channel_config_changed' ||
        typeof candidate.guildId === 'undefined' ||
        typeof candidate.channelId === 'undefined' ||
        typeof candidate.agentEnabled !== 'boolean'
      ) {
        return;
      }

      const guildId = String(candidate.guildId);
      const channelId = String(candidate.channelId);
      const cacheKey = `${CACHE_KEY_PREFIX}${guildId}:${channelId}`;

      this.cacheService.invalidate(cacheKey).catch(() => {
        this.logger.warn({ cacheKey }, 'Failed to invalidate agent config cache');
      });
    };

    this.eventPublisher.register(this.eventHandler);
  }

  dispose(): void {
    if (!this.eventHandler) {
      return;
    }
    this.eventPublisher.unregister(this.eventHandler);
    this.eventHandler = null;
  }
}
